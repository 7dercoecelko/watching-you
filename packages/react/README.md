# react-watching-you

<a href="https://www.npmjs.com/package/react-watching-you"><img src="https://badge.fury.io/js/react-watching-you.svg" alt="npm version" height="18"></a>

Watching-you is a javascript library for building animations that watch anything on DOM.

![watching your mouse](https://github.com/jj811208/watching-you/blob/main/static/1.gif)(watching your mouse)

![watching your input value](https://github.com/jj811208/watching-you/blob/main/static/2.gif)(watching your input value)

## Installation

### npm

```
npm install react-watching-you
```

### yarn

```
yarn add react-watching-you
```

## Quickstart

```javascript
import WatchingYou, { useWatchingYou } from 'react-watching-you';

// render props
const App = () => {
  return (
    <WatchingYou>
      <div>O</div>
    </WatchingYou>
  );
};

// hook
const App = () => {
  const [watchingYouWatcherProps] = useWatchingYou(watchingYouProps);
  return <div {...watchingYouWatcherProps}>O</div>;
};
```

## API Reference

<a href="https://jj811208.github.io/watching-you" target="_blank">https://jj811208.github.io/watching-you</a>
