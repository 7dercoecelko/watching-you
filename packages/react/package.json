{
  "name": "react-watching-you",
  "version": "0.0.1",
  "license": "ISC",
  "type": "module",
  "main": "./dist/react-watching-you.umd.cjs",
  "module": "./dist/react-watching-you.js",
  "types": "./dist/react/src/index.d.ts",
  "files": [
    "dist"
  ],
  "exports": {
    ".": {
      "import": "./dist/react-watching-you.js",
      "require": "./dist/react-watching-you.umd.cjs"
    }
  },
  "keywords": [
    "watching-you",
    "javascript",
    "animation",
    "react",
    "react-component"
  ],
  "scripts": {
    "start": "vite",
    "build": "tsc && vite build",
    "preview": "vite preview"
  },
  "devDependencies": {
    "@types/react": "^18.0.17",
    "@types/react-dom": "^18.0.6",
    "@vitejs/plugin-react": "^2.0.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "typescript": "^4.6.4",
    "vite": "^3.1.0",
    "vite-plugin-dts": "^1.4.1",
    "vite-tsconfig-paths": "^3.5.0"
  },
  "peerDependencies": {
    "react": "^16.8.0 || ^17 || ^18",
    "react-dom": "^16.8.0 || ^17 || ^18"
  },
  "repository": "https://github.com/jj811208/watching-you",
  "homepage": "https://github.com/jj811208/watching-you",
  "bugs": "https://github.com/jj811208/watching-you/issues",
  "author": "Ryan Chu <jj811208@gmail.com> (https://github.com/jj811208)"
}
