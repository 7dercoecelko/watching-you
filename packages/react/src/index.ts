import React, { useRef, useState, useEffect, useMemo } from 'react';
import WatchingYouBase, {
  WatchingYouRenderTransform,
  WatchingYouOptions,
} from '../../core/src/index';
import { concatObject } from './util';

interface WatchingYouReactHookProps
  extends Omit<WatchingYouOptions, 'render'> {
  active?: boolean;
}
interface WatchingYouReactHocProps extends WatchingYouReactHookProps {
  children: React.ReactNode;
}

const useWatchingYou = (props: WatchingYouReactHookProps = {}) => {
  const {
    target,
    targetType,
    power,
    rotatable,
    movable,
    active = true,
  } = props;
  const watcherRef = useRef(null);
  const [transform, setTransform] =
    useState<WatchingYouRenderTransform>({
      translate: { x: 0, y: 0 },
      rotate: 0,
    });
  const watchingYouRef = useRef(
    new WatchingYouBase({
      ...props,
      customRender: (newTransform) => {
        setTransform(newTransform);
      },
    }),
  );
  useEffect(() => {
    watchingYouRef.current.setTarget({
      target,
      targetType,
    });
  }, [target, targetType]);
  useEffect(() => {
    watchingYouRef.current.setPower(power);
  }, [power]);
  useEffect(() => {
    watchingYouRef.current.setRotatable(rotatable);
  }, [rotatable]);
  useEffect(() => {
    watchingYouRef.current.setMovable(movable);
  }, [movable]);
  useEffect(() => {
    watchingYouRef.current.setWatcher(
      watcherRef.current || undefined,
    );
  }, [watchingYouRef]);
  useEffect(() => {
    if (active) {
      watchingYouRef.current.start();
      return watchingYouRef.current.stop;
    } else {
      watchingYouRef.current.stop();
      return;
    }
  }, [active]);
  useEffect(() => {
    watchingYouRef.current.start();
    return watchingYouRef.current.stop;
  }, []);
  const watchingYouWatcherProps = useMemo(() => {
    return {
      ref: watcherRef,
      style: {
        transform: `translate(${transform.translate.x}px,${transform.translate.y}px) rotate(${transform.rotate}deg)`,
      },
    };
  }, [
    transform.translate.x,
    transform.translate.y,
    transform.rotate,
  ]);
  return [watchingYouWatcherProps, watchingYouRef.current] as const;
};
const WatchingYou: React.FC<WatchingYouReactHocProps> = (props) => {
  const { children, ...watchingYouProps } = props;
  const [watchingYouWatcherProps] = useWatchingYou(watchingYouProps);
  if (!React.isValidElement(children)) return null;
  return React.cloneElement(
    children,
    concatObject(children.props, {
      ref: watchingYouWatcherProps.ref,
      style: concatObject(
        children?.props?.style,
        watchingYouWatcherProps.style,
      ),
    }),
  );
};

export default React.memo(WatchingYou);
export { useWatchingYou };
