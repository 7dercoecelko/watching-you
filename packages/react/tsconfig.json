{
  "extends": "../../tsconfig",
  "jsx": "react-jsx",
  "compilerOptions": {
    "jsx": "react-jsx"
  },
  "include": [
    "src",
    "../core/src",
    "../core/demo"
  ],
  "references": [
    {
      "path": "../../tsconfig.node.json"
    }
  ]
}