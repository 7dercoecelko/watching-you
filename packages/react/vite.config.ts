import { defineConfig } from 'vite';
import { resolve } from 'path';
import react from '@vitejs/plugin-react';
import tsconfigPaths from 'vite-tsconfig-paths';

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [tsconfigPaths(), react()],
  build: {
    lib: {
      entry: resolve(__dirname, 'src/index.ts'),
      name: 'react-gazer',
      fileName: 'react-gazer',
    },
    rollupOptions: {
      external: ['react', 'react-dom'],
      output: {
        globals: {
          react: 'React',
          'react-dom': 'ReactDOM',
        },
      },
    },
  },
});
