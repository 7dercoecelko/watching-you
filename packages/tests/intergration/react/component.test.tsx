import React from 'react';
import { render, waitFor } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { ReactWatchingYou as WatchingYou } from '../../watchingYou';
import { mousemove } from '../../utils';

describe('Components', async () => {
  it('It should work', async () => {
    const watcher = 'O';
    const watchingYouElement = render(
      // @ts-ignore
      <WatchingYou>
        <div>{watcher}</div>
      </WatchingYou>,
    );
    await waitFor(() => mousemove(100, 100));
    expect(watchingYouElement.getByText(watcher)).toMatchSnapshot();
  });
});
