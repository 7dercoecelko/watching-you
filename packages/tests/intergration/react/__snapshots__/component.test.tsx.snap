// Vitest Snapshot v1

exports[`Components > It should work 1`] = `
<div
  style="transform: translate3d(39.7px,30.4px,0px) rotate(127.5deg);"
>
  O
</div>
`;

exports[`Components > It should work 2`] = `
<div
  style="transform: translate3d(39.7px,30.4px,0px) rotate(127.5deg);"
>
  O
</div>
`;
